File: video_11_minPathSumInTriangularGrid.java
import java.util.ArrayList;
class video_11_minPathSumInTriangularGrid {
//method 1: memoization
    // public static int[][] dp;
    // public static int solve(int row, int col, int n, ArrayList<ArrayList<Integer> >grid){
    //     if(row==n-1) return dp[n-1][col] = grid.get(n-1).get(col);

    //     if(dp[row][col] != -1) return dp[row][col];

    //     int down = grid.get(row).get(col)+solve(row+1, col, n, grid);
    //     int diag = grid.get(row).get(col)+solve(row+1, col+1, n, grid);

    //     return dp[row][col] = Math.min(down, diag);
    // }

    // public static int minimizeSum(int n, ArrayList<ArrayList<Integer> >grid) {
    //     // Code here
    //     dp = new int[n][n];
    //     for(int i = 0; i<n; i++){
    //         for(int j = 0; j<n; j++)
    //             dp[i][j] = -1;
    //     }
    //     return solve(0, 0, n, grid);
    // }

//method 2: tabulation
    // public static int minimizeSum(int n, ArrayList<ArrayList<Integer> >grid) {
    //     // Code here
    //     int dp[][] = new int[n][n];
    //     for(int i = 0; i<n; i++){
    //         for(int j = 0; j<n; j++)
    //             dp[i][j] = -1;
    //     }

    //     for(int i = 0; i<n; i++){
    //         dp[n-1][i] = grid.get(n-1).get(i);
    //     }

    //     for(int i = n-2; i>=0; i--){
    //         for(int j = i; j>=0; j--){
    //             int down = grid.get(i).get(j)+dp[i+1][j];
    //             int diag = grid.get(i).get(j)+dp[i+1][j+1];
    //             dp[i][j] = Math.min(down, diag);
    //         }
    //     }
    //     return dp[0][0];
    // }
    public static int minimizeSum(int n, ArrayList<ArrayList<Integer> >grid) {
        // Code here
        int[] front = new int[n]; 
        int[] cur = new int[n];

        for (int j = 0; j < n; j++) {
            front[j] = grid.get(n-1).get(j);
        }

        for(int i = n-2; i>=0; i--){
            for(int j = i; j>=0; j--){
                int down = grid.get(i).get(j)+front[j];
                int diag = grid.get(i).get(j)+front[j+1];
                cur[j] = Math.min(down, diag);
            }front = cur.clone();
        }return front[0];
    }
}

File: video_11_minPathSumInTriangularGrid.cpp
// #include<iostream>
// #include<bits/stdc++.h>
// using namespace std;

/*********************************************** MEMOIZATION ************************************************************** */
// int solve(int row, int col, vector<vector<int>>& triangle, int n, vector<vector<int>>& dp){
//     if(row == n-1) return dp[n-1][col]=triangle[n-1][col];
//     if(dp[row][col] != -1) return dp[row][col];

//     int d = triangle[row][col]+solve(row+1, col, triangle, n, dp);
//     int dg = triangle[row][col]+solve(row+1, col+1, triangle, n, dp);

//     return dp[row][col]=min(d, dg);    
// }
// int minimumPathSum(vector<vector<int>>& triangle, int n){
// 	// Write your code here.
//     vector<vector<int>> dp(n, vector<int>(n, -1));
//     return solve(0, 0, triangle, n, dp);
// }

/************************************************* TABULATION ************************************************************ */
// int minimumPathSum(vector<vector<int>>& triangle, int n){
// 	// Write your code here.
//     vector<vector<int>> dp(n, vector<int>(n, -1));
//     dp[n-1] = traingle[n-1];

//     for(int i = n-2; i>=0; i--){
//         for(int j = i; j>=0; j--){
//             int d = triangle[i][j]+dp[i+1][j];
//             int dg = triangle[i][j]+dp[i+1][j+1];
//             dp[i][j] = min(d, dg);
//         }
//     }
//     return dp[0][0];
// }

/************************************************** MAX SPACE OPTIMIZATION ********************************************** */
// int minimumPathSum(vector<vector<int>>& triangle, int n){
// 	// Write your code here.
//     vector<int> prev(n, 0);
//     vector<int> curr(n, 0);
//     prev = triangle[n-1];

//     for(int row = n-2; row>=0; row--){
//         for(int col = row; col>=0; col--){
//             int d = triangle[row][col]+prev[col];
//             int dg = triangle[row][col]+prev[col+1];
//             curr[col] = min(d, dg);
//         }
//         prev = curr;
//     }
//     return prev[0];
// }

/************************************************* MAIN ***************************************************************** */
// int main(){
//     int n;
//     cout<<"enter n: ";
//     cin>>n;
//     vector<vector<int>> triangle(n)
//     cout<<"enter elements in triangle: ";
//     for(int i = 0; i<n; i++){
//         for(int j = 0; j<=i; j++){
//             int element;
//             cin>>element;
//             triangle[i].push_back(element);
//         }
//     }
//     cout<<minimumPathSum(triangle, n);
// }